plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTable.java

```diff
@@ -0,0 +1,185 @@
+    {
+        assertQueryFails("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toTimestampString(1, 2),
+                "Cannot find a snapshot older than epoch time 1000");
+        assertQuery("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toTimestampString(v1Unixtime, 2), "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toTimestampString(v2Unixtime, 2), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testSelectTableWithEndLongTimestamp()
+    {
+        assertQueryFails("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toTimestampString(1, 8),
+                "Cannot find a snapshot older than epoch time 1000");
+        assertQuery("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toTimestampString(v1Unixtime, 8), "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toTimestampString(v2Unixtime, 8), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testSelectTableWithEndShortTimestampWithTimezone()
+    {
+        assertQueryFails("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toShortTimestampWithTimezoneString(1),
+                "Cannot find a snapshot older than epoch time 1000");
+        assertQuery("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toShortTimestampWithTimezoneString(v1Unixtime), "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toShortTimestampWithTimezoneString(v2Unixtime), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testSelectTableWithEndLongTimestampWithTimezone()
+    {
+        assertQueryFails("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toLongTimestampWithTimezoneString(1),
+                "Cannot find a snapshot older than epoch time 1000");
+        assertQuery("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toLongTimestampWithTimezoneString(v1Unixtime), "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_schema.test_table FOR TIMESTAMP AS OF " + toLongTimestampWithTimezoneString(v2Unixtime), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testEndVersionInTableNameAndForClauseShouldFail()
+    {
+        assertQueryFails("SELECT * FROM test_schema.\"test_table@" + v1SnapshotId + "\" FOR VERSION AS OF " + v1SnapshotId,
+                "Cannot specify end version both in table name and FOR clause");
+
+        assertQueryFails("SELECT * FROM test_schema.\"test_table@" + v1SnapshotId + "\" FOR TIMESTAMP AS OF " + toLongTimestampWithTimezoneString(v1Unixtime),
+                "Cannot specify end version both in table name and FOR clause");
+    }
+
+    private String toShortTimestampWithTimezoneString(long unixtime)
+    {
+        return "from_unixtime(" + unixtime + ")";
+    }
+
+    private String toLongTimestampWithTimezoneString(long unixtime)
```

Comment:
You are calling this method with "number of seconds since epoch". Seems wrong.

Reply:
I think `unixtime` is defined as that? 

> unixtime is the number of seconds since 1970-01-01 00:00:00 UTC.

in https://trino.io/docs/current/functions/datetime.html

Reply:
Not for `from_unix_timenanos`. 
![image](https://user-images.githubusercontent.com/890111/145961530-2a4cdd24-1e2a-4b60-8f8b-dc835a891250.png)

Reply:
I see, that makes sense, also explains why I had to time 1_000_000_000 above. Fixed the query.